Stats partial refactor and improvement

Piccions opened the pull request:
This PR improves the statistics subsystem:
- fixed the wrong statistics values (in particular the resident set size statistics)
- improved the performance of the GVT statistics reduction through vectorization and improvement of the buffer
- fixed the various statistics option (in particular "lp")
- signal SIGINT is properly handled, doing cleanup and flushing partial statistics
- cool formatting of the gvt and lps statistics files
- the ROOT-Sim run-time configuration is saved in the global statistics file as well
- improved general code quality of the module

alessandropellegrini committed:
Silence ar warning during make

Silence warning: ar: 'u' modifier ignored since 'D' is the default
This has to do with libtool. Different versions of libtool rely on
different macros, so we set the flags to the correct `cr` value
in both AR_FLAGS and ARFLAGS. AR_FLAGS is currently deprecated,
so it should go away at some point.

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Merge branch 'lighter-ult' into release-2.0.0

alessandropellegrini committed:
Merge branch 'develop' into release-2.0.0

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Merge pull request #116 from HPDCS/arch2

Reorganize src/arch/ subtree

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Fixing travis script

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Simplify timing subsystem

We directly rely on gettimeofday() to read the time, as it has
a very good performance on modern systems.

This also simplifies the configuration of the environment.

Fixes #98.

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Merge branch 'release-2.0.0' into develop

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

Commit:
Statistics code refresh and refactor.

The statistics module has seen a major restyling:

-the configuration is now being printed in the global file as well
-the reported resident set sizes, average and maximum, are now correct
-the lp files and the gvt files are now decently formatted in columns
-now we distinguish better between serial stats API and parallel ones
-the gvt buffer for the prints is far more efficient now
-improved general code quality

There are still some issues:

-STAT_GET_SIMTIME_ADVANCEMENT is an unused stats code
-STAT_GVT_ROUND_TIME is only partially implemented,
	that results in an horrible INFTY value being printed
-the code arguably needs more refactoring to eliminate some long functions
-documentation !!!

In core.c I'm introducing signal handling.
The feature is still in a development phase,
first of all I'll try to write a complete and sound cleanup function;
the current one, namely simulation_shutdown(), often glitches badly.

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccions@gmx.com>

Commit:
Fixed some missing mpi specific changes.

Commit:
Hot fix for MPI.

Introduced a workaround in gvt.c: will be fixed to a better state very soon.

Commit:
Continuing statistics refactor

Now when summing statistics gcc builtin vectors are used.
The option for statistics "lp" now is properly handled.
Improved code quality using foreach where possible.
Some other minor changes not worth listing here.

Still need some tests, especially on MPI.
This will be done when the develop branch will be fixed.

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccions@gmx.com>

Commit:
Fixed MPI statistics

-fixed a bug happning when multiple MPI processes share the same filesystem
-gvt round time statistics are now written only for relevant runs

Criticism and suggestions are always welcome

Signed-off-by: Andrea Piccione <piccions@gmx.com>

Commit:
Simple SIGINT handling, improved MPI prints

Sending SIGINT to a ROOT-Sim process now causes its graceful termination.
Fixed a regression in which slave processes printed the configuration and the termination message.

There's still the issue of the models printing on all the machines their configuration (if they do so from ARGP_SUCCESS).
I'm not sure either that termination of a kernel is broadcasted to all MPI kernels in a distributed setting.

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccions@gmx.com>

Commit:
Fixed GCC-8 warnings for snprintf()

The paths strings in statistics.c now use dynamically allocated buffers to comply with the overly zealous warning of gcc-8 regarding the possible truncation.

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccions@gmx.com>

alessandropellegrini commented on src/statistics/statistics.c:
```diff
@@ -242,23 +259,24 @@ static char *format_size(double size) {
+	f = thread_files[STAT_FILE_T_GVT][local_tid];
+	len = gvt_buf.pos;
+	for(i = 0; i < len; ++i){
+		fprintf(f, " %15lf    %15lf    %15u    %15u\n",
```
Watch out! While your implementation is nicer, the string buffer was done for performance reasons. Indeed, in the old code, this entailed a single interaction with the stdlib and a single interaction with the VFS to dump the content. Since this file can become very large (we reached GBs), letting the kernel optimize a single write could be more efficient.
Do you have any clue on whether this loop of printfs could affect performance?

alessandropellegrini commented on src/statistics/statistics.c:
```diff
@@ -533,114 +511,175 @@ void statistics_stop(int exit_code) {
-			if(exit_code == EXIT_SUCCESS) {
-				printf("\n--------- SIMULATION CORRECTLY COMPLETED ----------\n");
+			if(master_kernel()){
+				// TODO: do the statistics reduction between all the MPI kernels in another file
```
Any chance to have this implemented? :smile: 
Basically this entails having an `MPI_Reduce` in rank 0 (before checking if I am the master kernel) to compute the average of `system_wide_stats` coming from all ranks.
Given that we're dealing with a struct and we want to compute the average, we should define beforehand a new reduction operation using `MPI_Op_create`, maybe when initializing the statistics subsystem.

alessandropellegrini reviewed (commented):
That's a good refactor! Thanks!
Only two small comments raised in the code. The one on performance should be easy to sort out. The one on MPI reduction maybe requires a bit more tinkering...

Commit:
Improved gvt buffer writes on disk

Now the gvt buffer is flushed on a temporary line buffer to limit the calls to fprintf().
Still needs further testing and little tweaks!

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccions@gmx.com>

alessandropellegrini committed:
Sketching MPI reduction of statistics

The building blocks for the implementation of global reduction of
statistics are in place... good luck! :-)

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

Commit:
MPI-wide statistics redux, improved gvt stats buffer

Now, if running on MPI with more than one kernel, a global MPI statistics file is created and populated.

Now the gvt stats buffer is periodically flushed on a binary file: the actual formatted gvt stats file is created only at simulation termination.
This minimizes the frequency and the amount of disk writes during simulation.

Criticism and suggestions are always welcome!

Signed-off-by : Andrea Piccione <piccions@gmx.com>

Commit:
Little fixes to some MPI-wide stats.

GVT round time stats are now correctly reduced among MPI kernels.
Peak resident set sizes are summed up (it's clearly an overestimation of the real memory peak of the MPI run).

Criticism and suggestions are welcome!

Signed-off-by: Andrea Piccione <piccions@gmx.com>

alessandropellegrini committed:
Merge branch 'develop' into stats

Commit:
Fixed a bug and a regression in statistics.c

Now the MPI reduction of the simulation time speed statistic is correct.
Fixed a regression causing a segfault when the thread stopping the timer is different to the one reducing the stats.

Criticism and suggestions are welcome!

Signed-off-by: Andrea Piccione <piccions@gmx.com>

Commit:
Merge branch 'stats' of https://github.com/HPDCS/ROOT-Sim into stats
